ext/release-lead.rake

```diff
@@ -0,0 +1,116 @@
+  # Output the resulting `git describe` for that reference.
+  def git_describe_repo(name, url, sha_or_tag, where_to_clone, show_extra_commits)
+    Dir.chdir(where_to_clone) do
+      puts "Cloning #{name}..."
+      `git clone #{url} 2> /dev/null`
+      # `git clone --local '../../#{name}' 2> /dev/null`
+      # 0 for successful clone, 128 for already exists.
+      unless [0, 128].include? $?.exitstatus
+        next 'Could not find remote repository'
+      end
+      Dir.chdir(name) do
+        `git fetch 2> /dev/null`
+        `git checkout #{sha_or_tag} 2> /dev/null`
+        unless $?.exitstatus == 0
+          puts "Could not find reference #{sha_or_tag}"
+          exit
+        end
+        `git describe --tags`.strip +
+            if show_extra_commits
+              lines = `git log \`git describe --tags --abbrev=0\`..HEAD --oneline`.split("\n")
+              lines.map do |value|
+                if value.include?('[no-promote]')
+                  # Output these in red.
+                  "\e[31m#{value}\e[0m"
+                else
+                  value
+                end
+              end.join("\n  ")
+            end
+      end
+    end
+  end
+
+  def check_components(args, show_extra_commits = false)
+    result = {}
+    abort('Error: puppet_agent_branch argument is required') unless args.puppet_agent_branch
+
+    where_to_clone = File.join(File.dirname(__FILE__), 'pkg')
+    Dir.mkdir(where_to_clone) unless Dir.exists?(where_to_clone)
+
+    # Let's ensure puppet-agent is on the right branch.
+    `git fetch`
+    begin
+      `git checkout #{args.puppet_agent_branch} 2> /dev/null`
+
+      Dir.glob("./configs/components/*.json").each do |component|
+        json = JSON.parse(File.read(component))
+        url = json['url']
+        if url.nil?
+          name = json_name_to_component_name(component)
+          result[name] = 'No URL provided'
+          next
+        end
+        name = url_to_component_name(url)
+        ref = json['ref']
+        sha_or_tag = ref =~ /^refs\/tags/ ? ref.gsub('refs/tags/', '') : ref
+        result[name] = git_describe_repo(name, url, sha_or_tag, where_to_clone, show_extra_commits)
+      end
+
+      max_length = result.keys.map(&:length).max
+      puts "\n** Latest versions in #{args.puppet_agent_branch}:\n"
+      result.map do |name, value|
+        puts name.ljust(max_length + 1) + ': ' + value
+      end
+    ensure
+      `git checkout @{-1} 2> /dev/null`
+    end
+  end
+
+  # This task performs the following for each repository:
+  # - Looks up each component
+  # - Clones the component into the 'pkg' directory, if not done already
+  # - Checks out the git reference that's in the component's .json.
+  # - Outputs the `git describe` for that reference.
+  # This is useful for determining tag versions for an upcoming release.
+  desc "Output `git describe` for each component's git branch"
+  task :check_components_diff, [:puppet_agent_branch] do |t, args|
```

Comment:
Why not have an option to show the commits? Seems weird having two separate tasks. Also, I wonder if it'd be worthwhile to:
```
- Have an option to show only one component
- Have an option to show the summary of each commit (hence why the previous option might be useful)
```

If people really want to see the full diff, they can use the git UI.

Don't have strong feelings, just a thought.

Reply:
The shortcoming of adding an option is in how arguments are passed to rake... in this code, `:puppet_agent_branch` is the only argument passed in `[]`; using an option would mean invocation via `[5.5.x, <...?>]`. If arguments were named, I'd be fine with that, but it's not simple right now, so I split it into a separate task.

What do you mean by showing the summary of each commit, like the git commit message summary?

Reply:
Ah I see. So rake doesn't let you do option flags then? Something like `rake task --opt1=val1 ...`?

Yup, that's what I mean by summary. So something like:
```
<commit_title>:
<commit_summary>
<commit_title>:
<commit_summary>
```

would be useful.

Reply:
I think you have to do one or the other? I don't want to type out `puppet_agent_branch=...` for invocations if I can help it.

I see how commit title and description would be helpful, making it easier to hone in on what's changed in a certain component. I'll leave that as a possible improvement for the next time release lead duties come up.

Reply:
Cool, sounds good.